lib/compress/zstd_compress.c

```diff
@@ -2693,6 +2755,619 @@ static void ZSTD_confirmRepcodesAndEntropyTables(ZSTD_CCtx* zc)
+        assert(max <= defaultMax);
+        cSymbolTypeSizeEstimateInBits = max <= defaultMax
+                ? ZSTD_crossEntropyCost(defaultNorm, defaultNormLog, countWksp, max)
+                : ERROR(GENERIC);
+    } else if (type == set_rle) {
+        cSymbolTypeSizeEstimateInBits = 0;
+    } else if (type == set_compressed || type == set_repeat) {
+        cSymbolTypeSizeEstimateInBits = ZSTD_fseBitCost(fseCTable, countWksp, max);
+    }
+    if (ZSTD_isError(cSymbolTypeSizeEstimateInBits)) {
+        return nbSeq * 10;
+    }
+    while (ctp < ctEnd) {
+        if (additionalBits) cSymbolTypeSizeEstimateInBits += additionalBits[*ctp];
+        else cSymbolTypeSizeEstimateInBits += *ctp; /* for offset, offset code is also the number of additional bits */
+        ctp++;
+    }
+    return cSymbolTypeSizeEstimateInBits >> 3;
+}
+
+/* Returns the size estimate for the sequences section (header + content) of a block */
+static size_t ZSTD_estimateBlockSize_sequences(const BYTE* ofCodeTable,
+                                                  const BYTE* llCodeTable,
+                                                  const BYTE* mlCodeTable,
+                                                  size_t nbSeq,
+                                                  const ZSTD_fseCTables_t* fseTables,
+                                                  const ZSTD_fseCTablesMetadata_t* fseMetadata,
+                                                  void* workspace, size_t wkspSize,
+                                                  int writeEntropy)
+{
+    size_t sequencesSectionHeaderSize = 1 /* seqHead */ + 1 /* min seqSize size */ + (nbSeq >= 128) + (nbSeq >= LONGNBSEQ);
+    size_t cSeqSizeEstimate = 0;
+    cSeqSizeEstimate += ZSTD_estimateBlockSize_symbolType(fseMetadata->ofType, ofCodeTable, MaxOff,
+                                         nbSeq, fseTables->offcodeCTable, NULL,
+                                         OF_defaultNorm, OF_defaultNormLog, DefaultMaxOff,
+                                         workspace, wkspSize);
+    cSeqSizeEstimate += ZSTD_estimateBlockSize_symbolType(fseMetadata->llType, llCodeTable, MaxLL,
+                                         nbSeq, fseTables->litlengthCTable, LL_bits,
+                                         LL_defaultNorm, LL_defaultNormLog, MaxLL,
+                                         workspace, wkspSize);
+    cSeqSizeEstimate += ZSTD_estimateBlockSize_symbolType(fseMetadata->mlType, mlCodeTable, MaxML,
+                                         nbSeq, fseTables->matchlengthCTable, ML_bits,
+                                         ML_defaultNorm, ML_defaultNormLog, MaxML,
+                                         workspace, wkspSize);
+    if (writeEntropy) cSeqSizeEstimate += fseMetadata->fseTablesSize;
+    return cSeqSizeEstimate + sequencesSectionHeaderSize;
+}
+
+/* Returns the size estimate for a given stream of literals, of, ll, ml */
```

Comment:
Question:
do you have a debug mode
in which the estimation and the real size of the block are compared ?
If yes, do you have results ?

Reply:
I'll add this into the debuglogs. While developing this/eyeballing: the literals size estimate was typically almost exact (1-4 bytes off) while the sequences size estimate was a bit further off.

An excerpt of results from the beginning of `silesia.tar`:
```
../lib/compress/zstd_compress.c: Estimated size: 8108 actual size: 8158 
../lib/compress/zstd_compress.c: Estimated size: 6107 actual size: 6129 
../lib/compress/zstd_compress.c: Estimated size: 10062 actual size: 10058 
../lib/compress/zstd_compress.c: Estimated size: 19766 actual size: 19718 
../lib/compress/zstd_compress.c: Estimated size: 20583 actual size: 20464 
../lib/compress/zstd_compress.c: Estimated size: 22150 actual size: 22076 
../lib/compress/zstd_compress.c: Estimated size: 19845 actual size: 19756 
../lib/compress/zstd_compress.c: Estimated size: 21328 actual size: 21261 
../lib/compress/zstd_compress.c: Estimated size: 17800 actual size: 17706 
../lib/compress/zstd_compress.c: Estimated size: 19530 actual size: 19484
```